### PR TITLE
Mathesar 896 Change records api parameters to use column id instead of name

### LINKED ISSUE #896: Modify records API to use column id instead of column name
I'd like to be able to use the records API when all I know and care about is the column id for my columns.

## GET request params

The `order_by`, `group_count_by`, and `filters` params all expect references to columns to be given by name. For example:

- `order_by=[{"field":"name","direction":"asc"}]`
- `group_count_by=["name"]`
- `filters={"and":[{"field":"name","op":"eq","value":"Sean"}]}`

I'd like to the API to accept column references by `id`.

### As a breaking change

A simple way to address this need would be to break the API, no longer accepting column references by name

- `order_by=[{"field":"1","direction":"asc"}]`
- `group_count_by=["1"]`
- `filters={"and":[{"field":"1","op":"eq","value":"Sean"}]}`

### As a backwards compatible change

Another approach would be to accept _both_ references by name and by id. If I want to use the column id, I would do:

- `order_by=[{"field_id":"1","direction":"asc"}]`
- `group_count_by_ids=["1"]`
- `filters={"and":[{"field_id":"1","op":"eq","value":"Sean"}]}`

## GET Response

Currently the records API response look like:

```json
{
  "count": 2,
  "results": [
    {
      "id": 1,
      "first_name": "Bart",
      "last_name": "Simpson",
    },
    {
      "id": 2,
      "first_name": "Lisa",
      "last_name": "Simpson",
    }
  ]
}
```

To avoid relying on the column name, I'd like the response to give the column id as keys instead.

```json
{
  "count": 2,
  "results": [
    {
      "1": 1,
      "2": "Bart",
      "3": "Simpson",
    },
    {
      "1": 2,
      "2": "Lisa",
      "3": "Simpson",
    }
  ]
}
```

This would need to be a breaking change.

## POST and PATCH request and response 

Request and response bodies look like:

```json
{
  "name": "Sean"
}
```

I'd like to be able to update a record or add a new record by sending something like:

```json
{
  "1": "Sean"
}
```

## Additional context

- Related:
  - #839

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #896 

Change `/records/` api query parameters(`order_by`, `group_count_by`, and `filters`) to use column id instead of the column name.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
Blocked by Filtering API changes

**Comment by kgodey** [comment]
@silentninja Why is this blocked by the filtering API?

**Comment by seancolsen** [comment]
@silentninja I'd like to volunteer to add the front-end changes to this PR. Can you tag me when it's ready for front-end work?

**Comment by silentninja** [comment]
@kgodey Filtering API will be introducing breaking changes resulting in changes to the format of the json body of the `filter` query parameter, I had a talk with @dmos62 and he confirmed it. We might have to end up redoing this if we make the change now.

If this is a high priority issue, we can make the necessary changes now and redo it later along with filtering API

**Comment by kgodey** [comment]
@silentninja I think it's best to get this merged in, it's a fairly small change. @dmos62 can figure out how it affects the filtering API in his PR.

**Comment by pavish** [comment]
@silentninja I see that the frontend work for this is not yet done. I'm marking this PR as draft.

@seancolsen Would you like to do the frontend changes for this PR? I see that you have commented above.

**Comment by seancolsen** [comment]
@pavish Yep. I'll do this one since I opened the ticket for it. I'm just waiting on the go-ahead from @silentninja

**Comment by seancolsen** [comment]
Oh I see @silentninja marked it as "Ready for review". I'll interpret that as meaning it's ready for front end work, and I'll get started on this soon.

File: mathesar/api/pagination.py
Comment (kgodey) [inline mathesar/api/pagination.py:108]: @silentninja I haven't looked at this code in a while, this logic is hard to follow, can you add comments to explain what you're doing?

File: mathesar/api/pagination.py
Comment (kgodey) [inline mathesar/api/pagination.py:96]: Please don't name functions with single characters, be more verbose.

File: mathesar/api/viewsets/records.py
Comment (kgodey) [inline mathesar/api/viewsets/records.py:45]: Please remove print statement.

**Review by kgodey — CHANGES_REQUESTED** [review]
@silentninja My review is aimed at better understanding what's happening in this PR before I do a more thorough review, please see individual comments.

**Review by kgodey — APPROVED** [review]
This looks fine to me. Thanks @silentninja.

**Review by seancolsen — CHANGES_REQUESTED** [review]
@silentninja I see that using column ids now works for the GET request parameters. Great! But #896 also requires changes to the GET response which I'm not observing to be present in this PR. As well, the ticket requires changes to the POST and PATCH request and response which I'm assuming are also not present in this PR.

I request one of the following next steps:

- Edit the PR description, changing `Fixes #896` to `Related to #896` so that we leave the ticket open. We'd also need to make it clear within the ticket that some but not all of the work has been done.

    OR

- Add more changes to this PR to fully address the ticket.

**Comment by seancolsen** [comment]
@silentninja I'm going to keep myself assigned to this PR because, in addition to reviewing it, I'm also working on front end changes for it.

**Comment by silentninja** [comment]
@seancolsen Sorry I didn't notice the issue properly. Will push the additional fixes with this PR. I unassigned you, so that I can assign you back when I made the additional changes.

File: mathesar/api/db/viewsets/records.py
Comment (kgodey) [inline mathesar/api/db/viewsets/records.py:87]: Deserialization should be handled in the `.to_internal_value()` method of the serializer rather than in the viewset here. Also applies to other instances.

**Review by kgodey — CHANGES_REQUESTED** [review]
@silentninja Please see individual comments for requested changes.

**Comment by kgodey** [comment]
On the API level, this seems fine, my only comments were on implementation details. It should be ready for frontend changes, @seancolsen.

**Review by seancolsen — CHANGES_REQUESTED** [review]
The POST and PATCH request and response look good now. Thanks!

However, I have another small change to request...

- In the `master` branch, when I send a GET request without a `grouping` param, I recieve a response with `"grouping": null`. But in this branch I recieve:

    ```json5
    {
      // ...
      "grouping": {
        "columns": [],
        "mode": "distinct",
        "num_groups": null,
        "ranged": false,
        "groups": [
          {
            "count": 19,
            "first_value": {},
            "last_value": {},
            "result_indices": [
              0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18
            ]
          }
        ]
      },
      // ...
    }
    ```

    Can you change this to respond with `"grouping": null` like in `master`?

**Comment by kgodey** [comment]
To add to @seancolsen's comment, please create an API test for the grouping behavior so we catch future regressions.

**Comment by silentninja** [comment]
@kgodey I added  `grouping` field assertion check to [`mathesar/tests/api/test_record_api.py::test_record_list`](https://github.com/centerofci/mathesar/pull/1047/files#diff-b617b703f4686d1660e789a8de4c84a9d2cb347f5ec0c1261808528773507986)

File: mathesar/api/serializers/records.py
Comment (kgodey) [inline mathesar/api/serializers/records.py:48]: Why override this if you're only calling `super`? Please add a comment explaining it.

File: mathesar/api/db/viewsets/records.py
Comment (kgodey) [inline mathesar/api/db/viewsets/records.py:79]: I think this would be clearer as an `if` statement.

Comment (kgodey) [reply mathesar/api/db/viewsets/records.py:79]: Also applies below.

**Review by kgodey — APPROVED** [review]
Looks good, added some minor comments, please fix before merge. I don't think I need to review it again unless there are major changes.

File: mathesar/api/db/viewsets/records.py
Comment (silentninja) [reply mathesar/api/db/viewsets/records.py:79]: Most of the codebase seems to not use an `if` statement when calling `is_valid`, so I followed a similar approach but using an `if` statement is my preferred way of writing, would love to make that change in other places too along with this change to maintain consistency.

**Review by seancolsen — APPROVED** [review]
The API behavior here looks good now!

I'll get started on the front end changes. This is my highest priority right now, though it may take some time because I think the front end changes will require some refactoring.

**Comment by seancolsen** [comment]
I put front end changes in #1141 so that they can be reviewed separately and merged into this PR before we merge this PR into `master`

**Pull request merged**
